tests/functional/aws-node-sdk/test/versioning/replicationBucket.js

```diff
@@ -0,0 +1,80 @@
+const assert = require('assert');
+const async = require('async');
+
+const withV4 = require('../support/withV4');
+const BucketUtility = require('../../lib/utility/bucket-util');
+
+const bucketName = `versioning-bucket-${Date.now()}`;
+
+
+function checkError(err, code) {
+    assert.notEqual(err, null, 'Expected failure but got success');
+    assert.strictEqual(err.code, code);
+}
+
+function checkNoError(err) {
+    assert.ifError(err, `Expected success, got error ${JSON.stringify(err)}`);
+}
+
+function testVersioning(s3, versioningStatus, replicationStatus, cb) {
+    const versioningParams = { Bucket: bucketName,
+        VersioningConfiguration: { Status: versioningStatus } };
+    const replicationParams = {
+        Bucket: bucketName,
+        ReplicationConfiguration: {
+            Role: 'arn:aws:iam::123456789012:role/examplerole,' +
+            'arn:aws:iam::123456789012:role/examplerole',
+            Rules: [
+                {
+                    Destination: {
+                        Bucket: 'arn:aws:s3:::destinationbucket',
+                        StorageClass: 'STANDARD',
+                    },
+                    Prefix: '',
+                    Status: replicationStatus,
+                },
+            ],
+        },
+    };
+    async.waterfall([
+        cb => s3.putBucketReplication(replicationParams, e => cb(e)),
+        cb => s3.putBucketVersioning(versioningParams, e => cb(e)),
+    ], cb);
+}
+
+describe('Versioning on a replication source bucket', () => {
+    withV4(sigCfg => {
+        const bucketUtil = new BucketUtility('default', sigCfg);
+        const s3 = bucketUtil.s3;
+
+        beforeEach(done => {
+            async.waterfall([
+                cb => s3.createBucket({ Bucket: bucketName }, e => cb(e)),
+                cb => s3.putBucketVersioningPromise({
+                    Bucket: bucketName,
+                    VersioningConfiguration: {
+                        Status: 'Enabled',
+                    },
+                }, err => cb(err)),
+            ], done);
+        });
+
+        afterEach(done => s3.deleteBucket({ Bucket: bucketName }, done));
+
+        it('should not be able to disable versioning if replication enabled',
+        done => {
+            testVersioning(s3, 'Suspended', 'Enabled', err => {
+                checkError(err, 'InvalidBucketState');
+                done();
+            });
+        });
+
+        it('should be able to disable versioning if replication disabled',
+        done => {
+            testVersioning(s3, 'Suspended', 'Disabled', err => {
+                checkNoError(err);
+                done();
+            });
```

Comment:
You can just pass in done here, it will automatically fail if done is passed something that is not null or undefined. Also should apply to the other test.

```suggestion
            testVersioning(s3, 'Suspended', 'Disabled', done);
```

Reply:
This was done for readability, again you will see this across the test suite as mocha was not good at spitting out what went wrong. In the previous one we are testing for the error, so if we pass the error to done the test will fail, while we want it to pass when there is an error.

Reply:
oh whoops read that backwards, ignore this one lol